lalrpop/src/build/mod.rs

```diff
@@ -287,6 +287,54 @@ fn emit_uses<W:Write>(grammar: &r::Grammar,
     rust.write_uses("", grammar)
 }
 
+fn emit_interpreter(session: &Session,
+                         grammar: &r::Grammar)
+                         -> io::Result<Vec<u8>>
+{
+    let mut rust = RustWrite::new(vec![]);
+
+    try!(emit_uses(grammar, &mut rust));
+
+    if grammar.start_nonterminals.is_empty() {
+        println!("Error: no public symbols declared in grammar");
+        exit(1);
+    }
+
+    for (&user_nt, &start_nt) in &grammar.start_nonterminals {
+        // We generate these, so there should always be exactly 1
+        // production. Otherwise the LR(1) algorithm doesn't know
+        // where to stop!
+        assert_eq!(grammar.productions_for(start_nt).len(), 1);
+
+        log!(session, Verbose, "Building states for public nonterminal `{}`", user_nt);
+
+        let states = match lr1::build_states(&grammar, start_nt) {
+            Ok(states) => states,
+            Err(error) => {
+                let messages = lr1::report_error(&grammar, &error);
+                let _ = report_messages(messages);
+                exit(1) // FIXME -- propagate up instead of calling `exit`
+            }
+        };
+
+        try!(lr1::interpreter::compile(&grammar, user_nt, start_nt, &states, &mut rust));
```

Comment:
Ideally I think we would integrate this with `emit_recursive_ascent`, so that it's just something like `emit_parser` and we match here, since otherwise the code ought to be the same between the two cases I think